synapse_token_authenticator/token_authenticator.py

```diff
@@ -306,70 +310,191 @@ async def check_custom_flow(
+                        if username_type == "user_id"
+                        else None
+                    ),
+                ]
             )
-        except ValueError as e:
-            logger.info("Unrecognized token %s", e)
-            return None
-        except JWException as e:
-            logger.info("Invalid token %s", e)
-            return None
-        payload = json_decode(token.claims)
-        sub = payload["sub"]
-        if not isinstance(sub, str):
-            logger.info("user_id isn't a string")
-            return None
 
-        if "name" not in payload:
-            logger.info("No name claim in payload")
+            get_fq_uid_mb = if_not_none(lambda x: x.fq_uid_path)
+
+            fully_qualified_uid = all_list_elems_are_equal_return_the_elem(
+                [
+                    get_from_set(jwt_claims)(get_fq_uid_mb(config.jwt_validation)),
+                    get_from_set(introspection_claims)(
+                        get_fq_uid_mb(config.introspection_validation)
+                    ),
+                    username if username_type == "fq_uid" else None,
+                    (
+                        self.api.get_qualified_user_id(username)
+                        if username_type == "user_id"
+                        else None
+                    ),
+                    (
+                        self.api.get_qualified_user_id(username)
+                        if username_type == "localpart"
+                        else None
+                    ),
+                ]
+            )
+        except Exception as e:
+            logger.info(e)
             return None
 
-        user_id = UserID.from_string(user_id_str)
-        user_exists = await self.api.check_user_exists(user_id_str)
+        if localpart is None and fully_qualified_uid is None:
+            logger.info("No user id was provided")
+            return None
 
-        if not user_exists:
-            logger.info("User doesn't exist, registering them...")
-            await self.api.register_user(user_id.localpart)
+        if localpart is None:
+            localpart = UserID.from_string(fully_qualified_uid).localpart
 
-            # notification_access_token
-            headers = {}
-            if self.config.custom_flow.notification_access_token is not None:
-                headers = {
-                    b"Authorization": [
-                        b"Bearer " + self.config.custom_flow.notification_access_token
-                    ]
-                }
+        if fully_qualified_uid is None:
+            fully_qualified_uid = self.api.get_qualified_user_id(localpart)
 
-            await client.post_json_get_json(
-                self.config.custom_flow.notify_on_registration_uri,
-                {"token": login_dict["token"]},
-                headers=headers,
-            )
+        try:
+            get_displayname_mb = if_not_none(lambda x: x.displayname_path)
+            displayname = [
```

Comment:
Why is the displayname a list? 

When setting the displayname later synapse errors with status code 400 and message `'displayname' must be a string`. Referring to line 489:
```
await self.api._hs.get_profile_handler().set_displayname(
                requester=synapse.types.create_requester(user_id),
                target_user=user_id,
                by_admin=True,
                new_displayname=displayname,
            )
```

Reply:
Fixed